### PR TITLE
MTDSA-4117 Allowances, deductions and reliefs controller

### PR DESCRIPTION
**Pull request by jeremystone**

### CODE REVIEW

File: it/v1/endpoints/GetAllowancesDeductionsAndReliefsControllerISpec.scala [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            response.json shouldBe Json.toJson(expectedBody)
+          }
+        }
+
+        val input = Seq(
+          (BAD_REQUEST, "FORMAT_NINO", BAD_REQUEST, NinoFormatError),
+          (BAD_REQUEST, "FORMAT_CALC_ID", BAD_REQUEST, CalculationIdFormatError),
+          (NOT_FOUND, "MATCHING_RESOURCE_NOT_FOUND", NOT_FOUND, NotFoundError),
+          (INTERNAL_SERVER_ERROR, "SERVER_ERROR", INTERNAL_SERVER_ERROR, DownstreamError),
+          (SERVICE_UNAVAILABLE, "SERVICE_UNAVAILABLE", INTERNAL_SERVER_ERROR, DownstreamError)
```
Comment (BetaDraconis) [inline it/v1/endpoints/GetAllowancesDeductionsAndReliefsControllerISpec.scala]: This case isn't tested on any of the other endpoints. Does it need adding on the other ITs, or is it not required?

File: test/v1/controllers/GetAllowancesDeductionsAndReliefsControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  private val nino          = "AA123456A"
+  private val calcId        = "someCalcId"
+  private val correlationId = "X-123"
+
+  private val rawData     = GetCalculationRawData(nino, calcId)
+  private val requestData = GetCalculationRequest(Nino(nino), calcId)
+
+  private def uri      = s"/$nino/self-assessment/$calcId"
+  private def queryUri = "/input/uri?type=info&type=warning&type=error"
```
[the remark this answers is not in the document]
Comment (jeremystone) [reply test/v1/controllers/GetAllowancesDeductionsAndReliefsControllerSpec.scala]: Good spot. Removing from this and others (except get messages)

File: it/v1/endpoints/GetAllowancesDeductionsAndReliefsControllerISpec.scala (hunk 1 above)
Comment (jeremystone) [reply it/v1/endpoints/GetAllowancesDeductionsAndReliefsControllerISpec.scala]: These cases are in the tech spec. Will add to other ITs.

File: app/v1/models/response/getAllowancesDeductionsAndReliefs/AllowancesDeductionsAndReliefs.scala [hunk 3]
```diff
@@ -18,11 +18,19 @@ package v1.models.response.getAllowancesDeductionsAndReliefs
...
 import play.api.libs.json._
 
-case class GetAllowancesDeductionsAndReliefs(summary: CalculationSummary, detail: CalculationDetail)
+case class AllowancesDeductionsAndReliefs(summary: CalculationSummary, detail: CalculationDetail) {
 
-object GetAllowancesDeductionsAndReliefs {
-  implicit val reads: Reads[GetAllowancesDeductionsAndReliefs] =
-    (__ \ "allowancesDeductionsAndReliefs").read(Json.reads[GetAllowancesDeductionsAndReliefs])
+  def isEmpty: Boolean =
+    (summary.totalAllowancesAndDeductions, summary.totalReliefs) match {
```
[the remark this answers is not in the document]
Comment (jeremystone) [reply app/v1/models/response/getAllowancesDeductionsAndReliefs/AllowancesDeductionsAndReliefs.scala]: The backend looks like it just reads them as nullable as-is. 

In any case safer to be explicit without assumptions, and testable, I think.

[the remark this answers is not in the document]
Comment (BetaDraconis) [reply app/v1/models/response/getAllowancesDeductionsAndReliefs/AllowancesDeductionsAndReliefs.scala]: Ignore this.

**Pull request merged**